bam/reader.go

```diff
@@ -101,98 +99,121 @@ const (
 	AllVariableLengthData        // Omit sequence, quality and auxiliary data.
 )
 
-// Read returns the next sam.Record in the BAM stream. The sam.Record
-// returned will not contain the sequence, quality or auxiliary tag
-// data if CoreOnly(true) has been called.
-func (br *Reader) Read() (*sam.Record, error) {
-	if br.c != nil && vOffset(br.r.LastChunk().End) >= vOffset(br.c.End) {
-		return nil, io.EOF
-	}
+// buffer is light-weight read buffer.
```

Comment:
Can all the buffer implementation go below Reader (where the errReader etc were). The convention is that you lead with major moving parts and follow with helpers.